Implement `react-query` for the stake page

sam-potter opened the pull request:
PR for converting all the `redux` queries and mutations to `react-query` for the `/stake` page.

### Current conversion progress

- [x] Staking metrics
- [x] Rebase timer
- [x] Staking yield info
- [x] External stake pool queries
- [x] Staking token balances
- [x] Staking mutations
- [x] Test for correct balances, formatting, mutations, etc.

sam-potter committed:
Abstracted common Metric components

sam-potter committed:
Unified protocol metrics queries

sam-potter committed:
Remove unused query

sam-potter committed:
Preconnect to TheGraph external domain

sam-potter committed:
Added number formatting helper

sam-potter committed:
Preconnect to alchemy external domain

sam-potter committed:
Converted metrics to use `react-query`

sam-potter committed:
Converted yield metrics

sam-potter committed:
Update test snapshots

sam-potter committed:
Converted RebaseTimer

sam-potter committed:
Prefer `date-fns` library for formatting time

sam-potter committed:
Move RebaseTimer folder

sam-potter committed:
Update import

sam-potter committed:
Update import

sam-potter committed:
Abstracted components

sam-potter committed:
Revert `date-fns` time formatting changes

sam-potter committed:
Updated import

0xJem commented:
Hi @sam-potter , this looks very promising!

Do you have any benchmarks that you can share? e.g. a lighthouse score and/or performance report to demonstrate a quantifiable difference.

sam-potter committed:
Fixed infinite network calls

sam-potter committed:
Moved `ExternalStakePool` file

sam-potter committed:
Converted external stake pools

sam-potter committed:
Updated test snapshots

sam-potter committed:
Merge branch 'develop' into react-query/stake-view

sam-potter commented:
Great idea @0xJem, i'll do a couple benchmarks once all the queries have been converted over.

sam-potter committed:
Turned off all default refetches

sam-potter committed:
Converted stake balances

sam-potter committed:
Merge branch 'develop' into react-query/stake-view

sam-potter commented:
Latest merge includes changes in #1409 and #1407

sam-potter committed:
Force specifying network explicitly for contracts

sam-potter commented:
@0xJem There is a slight improvement to the largest contentful paint and the time to interactive metrics. There should also be an improvement in the speed index, except because the version on develop takes so long to load the external stake pool data, Lighthouse incorrectly assumes the loading state as the final state.

Imo, I think these metrics undervalue the perceivable improvement in speed though.

![screenshot](https://user-images.githubusercontent.com/15123145/152619348-ba00f243-3831-43b4-8b84-53d06a692ea9.png)

sam-potter committed:
Force specifying provider explicitly for contracts

sam-potter committed:
Merge branch 'develop' into react-query/stake-view

sam-potter committed:
Abstracted component

sam-potter committed:
Made `address` an invariant

sam-potter committed:
Abstracted `StakeInputArea`

sam-potter committed:
Merge branch 'develop' into react-query/stake-view

sam-potter committed:
Remove test

sam-potter committed:
Refactor

0xJem commented:
I ran a side-by-side test and this branch was MUCH quicker. This is amazing work!

https://user-images.githubusercontent.com/91760036/152936588-761adb6c-109e-44e6-8a2b-911ac0f1430d.mp4

sam-potter committed:
Converted mutations

sam-potter commented:
Just finished converting the mutations.

Realised I had accidentally made a UX change: in this branch, when you are unstaking from gohm the amount you enter into the input box is denominated in gohm. As opposed to in prod, where you denominate how much gohm you want to unstake in ohm.

I haven't immediately fixed it because I wanted to get some thoughts on this. Imo, inputting the gohm amount is more intuitive because that's the amount of gohm a user is unstaking, and if they wanted to unstake all their gohm they only need to copy and paste the balance in their wallet, rather than working out the conversion rate to precise decimal places.

Would love to hear any elses thoughts on whether a user should input gohm or ohm in that situation.

sam-potter commented:
Also for testing, unfortunately covalent doesn't support rinkeby :(( So you'll have to comment out the line in `useStakeMutation` that prevents staking/unstaking an amount greater than users balance.

sam-potter commented:
Fixes #1302

sam-potter committed:
Fix string interpolation bug

sam-potter committed:
Bug fixes

sam-potter committed:
Total balance now shown in sOHM

sam-potter committed:
Merge branch 'develop' into react-query/stake-view

sam-potter committed:
Fixed more string interpolation translation bugs

0xJem commented:
> Fixes #1302

Can you elaborate? I had also addressed this in #1414 (along with a race condition) and refactored the staking page to be much more easy to maintain.

sam-potter committed:
Fixed math underflow bug

sam-potter commented:
> > Fixes #1302
> 
> Can you elaborate? I had also addressed this in #1414 (along with a race condition) and refactored the staking page to be much more easy to maintain.

I noticed your fix yesterday (as well as the work @alexleventer had done) and thank you both for that! Since all of those improvements you had done were for the staking queries/mutations with redux, it was annoying timing with this conversion to react-query stuff happening simultaneously.

I ended taking huge inspiration from #1414 (especially the comments) to convert everything over, and basically now this PR fully fixes the issue in #1302.

Except I'm not sure about this race condition, so I'll pm you to learn more about that.

sam-potter committed:
Fix possible bug

nabioz commented:
> I haven't immediately fixed it because I wanted to get some thoughts on this. Imo, inputting the gohm amount is more intuitive because that's the amount of gohm a user is unstaking, and if they wanted to unstake all their gohm they only need to copy and paste the balance in their wallet, rather than working out the conversion rate to precise decimal places.
> 
> Would love to hear any elses thoughts on whether a user should input gohm or ohm in that situation.

I agree with the gohm change as input. I have heard multiple complaints about this as well so I think it's safe to say we are not alone.

nabioz commented:
I'm unable to get stuff loading. Is there specific configuration for .env for this to work? 
<img width="847" alt="image" src="https://user-images.githubusercontent.com/92614968/153222130-38210867-6083-4b76-b527-9d5d53bb8cd8.png">

sam-potter committed:
Rename contracts to be more explicit

sam-potter commented:
@mrunbanked yep you'll need all the provider keys setup in your `.env`. I'll pm that to you asap.

0xJem commented:
> I noticed your fix yesterday (as well as the work @alexleventer had done) and thank you both for that! Since all of those improvements you had done were for the staking queries/mutations with redux, it was annoying timing with this conversion to react-query stuff happening simultaneously.
> 
> I ended taking huge inspiration from #1414 (especially the comments) to convert everything over, and basically now this PR fully fixes the issue in #1302.
> 
> Except I'm not sure about this race condition, so I'll pm you to learn more about that.

1414 has a few changes:
1. Cleans up/simplifies the JSX code so there are fewer inline conditionals
    - I think you've achieved this with splitting into components
2. Fixes the approval status not being retained (#1302)
    - This seems to be fixed in the current Fleek deployment
3. Fixes the CTA button showing "approve" after the approval transaction is finished (#1451)
    - This issue still exists (I just tested it)
    - You can fix it by porting over my changes to StakeThunk: https://github.com/OlympusDAO/olympus-frontend/blob/60c321615241f8478485239cabfff0ae6a54255d/src/slices/StakeThunk.ts#L145

I can close the 1414 PR after that!

sam-potter committed:
Fixed button temporarily showing 'Approve'

sam-potter commented:
> This issue still exists (I just tested it)

Just implemented a fix for #1451, let me know if that's good.

0xJem commented:
> > This issue still exists (I just tested it)
> 
> Just implemented a fix for #1451, let me know if that's good.

Tested this on the Fleek build and the issue is fixed. 👍🏻

nabioz reviewed (approved):
Looks good to me. If you can remove v1-stake and any redux Stake related stuff and it works, we can definitely merge this! @appleseed-iii can you confirm that we are ok to remove v1-stake page? There seems to be no UI access to it or a use-case at this point with no OHMv1 liquidity

0xJem commented:
There are some conflicts with `develop` too

nabioz commented on src/components/TokenAllowanceGuard/hooks/useApproveToken.ts:
```diff
@@ -0,0 +1,37 @@
+      if (!token) throw new Error("Token doesn't exist on current network. Please switch networks.");
+      if (!contractAddress) throw new Error("Contract doesn't exist on current network. Please switch networks.");
+
+      const transaction = await token.approve(contractAddress, CONTRACT_APPROVAL_AMOUNT);
```
for approval amounts we can use ethers.constant.MAXUINT (something like that), to make sure we cover all decimal cases @sam-potter

sam-potter replied:
Done

sam-potter committed:
Prefer `MaxUint256`

appleseed-iii commented:
> Looks good to me. If you can remove v1-stake and any redux Stake related stuff and it works, we can definitely merge this! @appleseed-iii can you confirm that we are ok to remove v1-stake page? There seems to be no UI access to it or a use-case at this point with no OHMv1 liquidity

I'm not sure that we can remove `v1-stake` don't we still have some unmigrated wallets?

appleseed-iii commented on src/views/Stake/components/ExternalStakePools/hooks/useStakePoolTVL.ts:
```diff
@@ -0,0 +1,52 @@
+
+  // Get dependent data in parallel
+  const { data: gohmPrice } = useGohmPrice();
+  const reserves = useDependentQuery("reserves", () => contract.getReserves());
+  const firstTokenAddress = useDependentQuery("firstTokenAddress", () => contract.token0());
+  const poolTokenSupply = useDependentQuery("poolTokenSupply", () => contract.totalSupply());
+  const stakedBalance = useDependentQuery("stakedBalance", () => contract.balanceOf(pool.masterchef));
+  const nonGohmTokenPrice = useDependentQuery("nonGohmTokenPrice", () => getTokenPrice(pool.pairGecko));
```
wow, I hadn't seen `dependent queries` before. nice

sam-potter replied:
Cleanest way I could come up with for downloading all the other data that a query needs. And ensures requests are done in parallel, as soon as we know we need that data.

appleseed-iii reviewed (commented):
ok, going to run back through this one more time, but first pass looks good.

sam-potter committed:
Merge branch 'develop' into react-query/stake-view

nabioz commented:
> > Looks good to me. If you can remove v1-stake and any redux Stake related stuff and it works, we can definitely merge this! @appleseed-iii can you confirm that we are ok to remove v1-stake page? There seems to be no UI access to it or a use-case at this point with no OHMv1 liquidity
> 
> I'm not sure that we can remove `v1-stake` don't we still have some unmigrated wallets?

v1-stake page is not reachable through the UI. We will still keep the migration UI

sam-potter committed:
Improve reference stability

sam-potter committed:
Simplify static contract API

sam-potter committed:
Source balances from on-chain

sam-potter committed:
Merge branch 'develop' into react-query/stake-view

appleseed-iii commented:
just dropping this note. my current hang up is not being able to get testnet balances to load. I don't think we want to merge without that. 
I'm also going through & cataloguing network requests here & want to make sure we are replacing them all: https://www.notion.so/olympusdao/Stake-View-Original-Network-Requests-898bc24a443e4a64952d2d777162b226

sam-potter committed:
Remove unnecessary typings

sam-potter committed:
Separated staking mutations

sam-potter committed:
Balances refactor

sam-potter committed:
Improve testing environment

sam-potter committed:
Merge branch 'develop' into react-query/stake-view

sam-potter committed:
Update test snapshots

sam-potter committed:
Remove non-null values from query key

sam-potter commented:
> my current hang up is not being able to get testnet balances to load

Fixed now @appleseed-iii 

And watching them update in realtime, seriously satisfying

https://user-images.githubusercontent.com/15123145/154666468-0d66d89d-105e-44d2-b84a-eebf466ea060.mov

sam-potter committed:
Include gohm balance in calculation

sam-potter committed:
Merge branch 'develop' into react-query/stake-view

appleseed-iii commented on src/hooks/useTestableNetworks.ts:
```diff
@@ -0,0 +1,19 @@
+  const { networkId } = useWeb3Context();
+
+  return {
+    MAINNET: getTestnet(NetworkId.MAINNET, NetworkId.TESTNET_RINKEBY, networkId),
```
I'll add as a TODO for a separate PR to add `FANTOM` & `AVAX` testnets

appleseed-iii reviewed (approved):
verified:
- [x] useApproveToken
- [x] `src/constants/addresses` changes
- [x] `usePendingTransactions`
- [x] `useStakingRebaseRate`
- [x] migration modal prompt?

todo in separate pr:
- [ ] rework wallet to use `useBalance`
- [ ] add `FANTOM` & `AVAX` testnets to `useTestableNetworks`
- [ ] merge in https://github.com/OlympusDAO/olympus-frontend/pull/1476
- [ ] remove extra redux queries https://www.notion.so/olympusdao/Stake-View-Original-Network-Requests-898bc24a443e4a64952d2d777162b226
